src/ucp/proto/proto_init.c

```diff
@@ -196,180 +115,251 @@ ucp_proto_perf_envelope_make(const ucs_linear_func_t *funcs, uint64_t funcs_num,
-        perf_node = ucp_proto_perf_node_new_data("rcache lookup", "");
-
-        *memreg_time = ucs_linear_func_make(context->config.ext.rcache_overhead,
-                                            0);
-
-        ucp_proto_perf_node_add_data(perf_node, "lookup", *memreg_time);
-
-        goto out;
+        perf_factors[cpu_factor_id] =
+                ucs_linear_func_make(context->config.ext.rcache_overhead, 0);
+        return ucp_proto_perf_add_funcs(perf, range_start, range_end,
+                                        perf_factors, NULL, "rcache lookup",
+                                        "");
     }
 
-    perf_node = ucp_proto_perf_node_new_data("mem reg", "");
+    reg_perf_node = ucp_proto_perf_node_new_data("mem reg", "");
 
     /* Go over all memory domains */
     ucs_for_each_bit(md_index, reg_md_map) {
         md_attr = &context->tl_mds[md_index].attr;
         md_name = context->tl_mds[md_index].rsc.md_name;
-        ucs_linear_func_add_inplace(memreg_time, md_attr->reg_cost);
-        ucs_trace("md %s reg_cost: " UCP_PROTO_PERF_FUNC_FMT, md_name,
+        ucs_linear_func_add_inplace(&perf_factors[cpu_factor_id],
+                                    md_attr->reg_cost);
+        ucs_trace("md %s reg: " UCP_PROTO_PERF_FUNC_FMT, md_name,
                   UCP_PROTO_PERF_FUNC_ARG(&md_attr->reg_cost));
-
-        ucp_proto_perf_node_add_data(perf_node, md_name, md_attr->reg_cost);
+        ucp_proto_perf_node_add_data(reg_perf_node, md_name, md_attr->reg_cost);
     }
 
     if (!ucs_is_pow2(reg_md_map)) {
         /* Multiple memory domains */
-        ucp_proto_perf_node_add_data(perf_node, "total", *memreg_time);
+        ucp_proto_perf_node_add_data(reg_perf_node, "total",
+                                     perf_factors[cpu_factor_id]);
     }
 
-out:
-    *perf_node_p = perf_node;
+    status = ucp_proto_perf_add_funcs(perf, range_start, range_end,
+                                      perf_factors, reg_perf_node,
+                                      perf_node_name, "%u mds",
+                                      ucs_popcount(reg_md_map));
+    ucp_proto_perf_node_deref(&reg_perf_node);
+    return status;
+}
+
+static ucp_proto_perf_factor_id_t
+ucp_proto_buffer_copy_get_cpu_factor_id(int local) {
+    return (local) ? UCP_PROTO_PERF_FACTOR_LOCAL_CPU :
+                     UCP_PROTO_PERF_FACTOR_REMOTE_CPU;
+}
+static ucp_proto_perf_factor_id_t
+ucp_proto_buffer_copy_get_factor_id(ucs_memory_type_t local_mem_type,
+                                    ucs_memory_type_t remote_mem_type,
+                                    uct_ep_operation_t memtype_op,
+                                    int local)
+{
+    int h2h      = UCP_MEM_IS_HOST(local_mem_type) &&
+                   UCP_MEM_IS_HOST(remote_mem_type);
+    /* RNDV mtype protocols which do async copy set ZCOPY as `memtype_op`
+     * while eager procols that imply blocking copy set SHORT */
+    int blocking = (memtype_op == UCT_EP_OP_GET_SHORT) ||
```

Comment:
there is the same check on line 437 in this file. Maybe create an inline function for it?

Reply:
These checks are not the same. Check on 437 is:
```c
    if ((memtype_op == UCT_EP_OP_GET_SHORT) ||
        (memtype_op == UCT_EP_OP_GET_ZCOPY)) {
```

And it check whether operation PUT or GET, while this check defines whether operation blocking or not